### PR TITLE
Add `tf.SparseTensor` and `tf.IndexedSlices` support to a number of T…

### PR DESCRIPTION
**Pull request by hertschuh**
…ensorFlow ops.

The following element-wise unary ops now support `tf.SparseTensor` and `tf.IndexedSlices`. The output is of the same type as the input.
- abs
- absolute
- arcsin
- arcsinh
- arctan
- arctanh
- ceil
- conj
- conjugate
- copy
- expm1
- floor
- imag
- log1p
- negative
- real
- round
- sign
- sin
- sinh
- sqrt
- square
- tan
- tanh

The following element-wise unary ops now support `tf.SparseTensor` and `tf.IndexedSlices`. The output is dense.
- arccos
- arccosh
- cos
- cosh
- exp
- log
- log10
- log2
- reciprocal
   
The following element-wise binary ops now support `tf.SparseTensor` and `tf.IndexedSlices`. The output type depends on the two inputs and the op.
- add (already supported `tf.SparseTensor`)
- subtract (already supported `tf.SparseTensor`)
- maximum (already supported `tf.SparseTensor`)
- minimum (already supported `tf.SparseTensor`)
- multiply (already supported `tf.SparseTensor`)
- mod
- divide
- true_divide
- floor_divide

The following reduction op now supports `tf.IndexedSlices`. The output is an `tf.IndexedSlices` unless dimension 0 is reduced or the rank of the output is 1 or less.
- mean

This is in preparation for supporting sparse gradients in optimizers.

### CODE REVIEW

File: keras/backend/tensorflow/numpy.py [hunk 1]
```diff
@@ -10,17 +9,17 @@
 from keras.backend import config
 from keras.backend import standardize_dtype
 from keras.backend.common import dtypes
+from keras.backend.tensorflow import sparse
 from keras.backend.tensorflow.core import convert_to_tensor
 
 
+@sparse.element_wise_binary_on_union(tf.sparse.add)
```
Comment (fchollet) [inline keras/backend/tensorflow/numpy.py:16]: Let's apply the following replacements:

`element_wise` -> `elementwise` (everywhere)
`on_union` -> just `union` (to be consistent with `intersection`)

Comment (fchollet) [reply keras/backend/tensorflow/numpy.py:16]: Question: is this new system really better than `tf.sparse.add`? Does it have performance implications? The previous "`if` + use `tf.sparse`" setup was very readable and easy to maintain. The new system is fairly high in abstraction and complexity. Are the benefits commensurate?

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

**Comment by fchollet** [comment]
To note, `cond` ops can be quite expensive. We should check what performance looks like.

File: keras/backend/tensorflow/numpy.py (hunk 1 above)
Comment (hertschuh) [reply keras/backend/tensorflow/numpy.py:16]: Will do the replacements

> is this new system really better than `tf.sparse.add`?

So there are 2 times 3 cases: `SparseTensor` / `IndexedSlices` and **union** (add, maximum...), **intersection** (multiply), **division** (divide, true_divide, mod, floor_divide). `tf.sparse` happens to mostly support the "union" operators out of the box, so it was very little code. But it didn't support the other use cases and `IndexedSlices` is not supported for any operation. There is a lot of code for the 5 other use cases (for instance, this CL moved the long chunk of code to get `multiply` working for sparse from `numpy.py` to `sparse.py`).

One thing we could do differently is have the `elementwise_binary_union` decorator only add `IndexedSlices` support and not `SparseTensor`s to have the `tf.sparse.add` code inline. I initially had it this way, but then I decided to make it look consistent.

> Does it have performance implications?

For `tf.sparse.add` specifically, the only performance impact is the call to the decorator, which we need anyway for the `IndexedSlices` support. It does use `tf.sparse.add` directly, this is not re-implemented.

> The new system is fairly high in abstraction and complexity.

I agree. After all was done to handle all the possible combinations, it was a lot more complex than initially expected.

> Are the benefits commensurate?

Well the benefit is that with one decorator, you make an op `SparseTensor` and `IndexedSlices` compatible. And it's self documenting, you know which ops support `SparseTensor` and `IndexedSlices` (we could generate documentation based on the decorator). I could have decorated more ops actually (e.g. logical operations).

I want to look at JAX and see what approach makes sense for JAX. Maybe it will be similar, maybe it will make me rethink this whole thing.

**Comment by hertschuh** [comment]
> To note, `cond` ops can be quite expensive. We should check what performance looks like.

Interesting. Some of the `cond`s are optimizations (shortcut when indices are the same, which is a very common use case with sparse gradients). But some are workarounds for TF bugs. For instance, if after computing the intersection nothing is else (empty `SparseTensor` with no indices), a bunch of trivial operations fails (e.g. reshape) even though they're perfectly valid, so I had to special case them.

File: keras/backend/tensorflow/numpy.py (hunk 1 above)
Comment (fchollet) [reply keras/backend/tensorflow/numpy.py:16]: Thanks for the clarifications, this all sounds reasonable. Please let me know what you think about the JAX situation, and then we can finalize the design.

Comment (hertschuh) [reply keras/backend/tensorflow/numpy.py:16]: Thanks!

One other thing is that implementing support in the ops directly adds complexity. For instance, in practice, this union / intersection business is never needed for `IndexedSlices`. In optimizers the only times `IndexedSlices` are combined, they always have the same indices. But in the ops, we can't just make this assumption.

**Comment by hertschuh** [comment]
> > To note, `cond` ops can be quite expensive. We should check what performance looks like.
> 
> Interesting. Some of the `cond`s are optimizations (shortcut when indices are the same, which is a very common use case with sparse gradients). But some are workarounds for TF bugs. For instance, if after computing the intersection nothing is else (empty `SparseTensor` with no indices), a bunch of trivial operations fails (e.g. reshape) even though they're perfectly valid, so I had to special case them.

Looking into this a bit more, I found a way to check that the indices are the same at trace time. So, in the case of optimizers for instance, the `cond`s should no longer be in the graph because when we combine two `IndexedSlices`, they come from the same gradient and have the same indices.

**Review by fchollet — APPROVED** [review]
LGTM, thank you for the contribution!
